### PR TITLE
Test pathological cases in parsing code

### PR DESCRIPTION
**Pull request by snim2**
There is a comment on PR #24 that "all the tests pass, but I don't know if the tests touch all of these code paths". This PR is an attempt to add tests for error types defined in `ast::ParseError`, and to simplify that enumeration.

Can be squashed into one commit.

### CODE REVIEW

File: tests/test_ast.rs [hunk 1]
```diff
@@ -234,3 +234,105 @@ fn test_hello_java() {
...
+
+#[test]
+fn test_non_existant_input() {
+    let lex = Path::new("grammars/calc.l");
+    let yacc = Path::new("grammars/calc.y");
+    let ast_result = parse_file("nosuchfileexists.calc", lex, yacc);
+    assert!(ast_result.is_err());
+    match ast_result {
+        Err(ParseError::FileNotFound(s)) => {
+            if s == String::from("nosuchfileexists.calc") {
```
Comment (ltratt) [inline tests/test_ast.rs]: Why not just replace this whole `if` with `assert_eq!(s, "nosuchfileexists.calc")` or similar?

File: tests/test_ast.rs [hunk 2]
```diff
@@ -234,3 +234,105 @@ fn test_hello_java() {
...
+
+#[test]
+fn test_non_existant_lex() {
+    let lex = Path::new("grammars/nosuchfileexists.l");
+    let yacc = Path::new("grammars/calc.y");
+    let ast_result = parse_file("tets/one.calc", lex, yacc);
+    assert!(ast_result.is_err());
+    match ast_result {
+        Err(ParseError::FileNotFound(s)) => {
+            if s == String::from("grammars/nosuchfileexists.l") {
```
Comment (ltratt) [inline tests/test_ast.rs]: Same comment as above.

File: tests/test_ast.rs [hunk 3]
```diff
@@ -234,3 +234,105 @@ fn test_hello_java() {
...
+        _ => assert!(false),
+    }
+}
+
+#[test]
+fn test_lexical_err() {
+    let lex = Path::new("grammars/calc.l");
+    let yacc = Path::new("grammars/calc.y");
+    let ast_result = parse_file("tests/calc_lexical_err.calc", lex, yacc);
+    assert!(ast_result.is_err());
```
Comment (ltratt) [inline tests/test_ast.rs]: This line is redundant given the `match`.

**Comment by snim2** [comment]
OK, ready to squash?

**Comment by ltratt** [comment]
In general, I tend to think `assert!(false)` is less clear than `panic!`. Any thoughts?

**Comment by snim2** [comment]
Well, in a unit test I would expect a test to fail, rather than to cause a panic. If you have lots of failing tests you may still want to run the whole suite to completion.

**Comment by ltratt** [comment]
`panic!` doesn't stop the test suite I'm pretty sure (maybe try to be sure!).

**Comment by snim2** [comment]
It seems you are right about `panic!`, change pushed... Good to squash now?

**Comment by ltratt** [comment]
Yep, looks good to me!

**Comment by ltratt** [comment]
Done?

**Comment by snim2** [comment]
Oops, yes!

**Pull request merged**